Good and bad examples (#3)
silverstripe-terraformers/accessibility-guidelines: docs/getting-started.md

# Getting started with accessibility

<a href="https://github.com/silverstripeltd/project-skeleton/blob/master/docs/accessibility.md
">accessibility.md</a> in project skeleton is a good starting point for checking whether your feature
has met accessibility guidelines.

The docs include a quick checklist as well as a more in-depth checklist should you choose to go with a more thorough check.

## The good vs the bad
Below you can find a list of good vs bad examples with what was mentioned in the checklist to make it more obvious as to why we should/shouldn't do something.

> [[toc]]

Reviewer: 🤯 nice

**Decorative image** doesn't add information to the content of the page.
Images may be decorative when they are:
1. Visual styling (borders, corners.etc)
2. Supplementary to surrounding text (improving the clickable area or only for illustrative purpose)

#### :tada: GOOD Decorative image example:
```html{2}
<a href="crocuspage.html">
	<img src="crocus.jpg" alt="">
	<strong> Crocus bulbs</strong>
</a>
```
You can also use `role="presentation"`, but it's not as well supported as the empty alt attribute
#### :lady_beetle: BAD decorative image example:
```html
<a href="crocuspage.html">
	<img src="crocus.jpg" alt="crocus bulbs">
	<strong> Crocus bulbs</strong>
</a>
```
::: tip
- This will result in the screen reader reading the `crocus bulbs` twice, `alt` is not needed to describe what's already been described by the surrounding text. **Only use alt with contextual image**
:::

**Contextual image** is the opposite and will represent concepts/information(i.e. Images conveying an impression or supplement other information)
#### :tada: GOOD contextual Image
```html
<img src="cap.png" alt="Push the cap down and turn it counter-clockwise (from right to left)">
```
:::warning

> Without adding the alt tag for contextual image, it could result in people using screen reader technologies missing out on important information.

Reviewer: ```suggestion
Without adding the alt tag for a contextual image, it could result in people using screen reader technologies missing out on important information.
```

:::

#### :lady_beetle: BAD contextual image example:
```html
<img src="cap.png" alt="instruction picture for how to open this special cap">
```
::: tip

> Alt should also be meaningful and describe the image accurately, it's especially important if the image is supplying information

Reviewer: ```suggestion
The `alt` attribute should also be meaningful and describe the image accurately, it's especially important if the image is supplying information
```

    c1.3,1.7,1.9,3,2.1,4.7C19.8,33,28.2,33,28.2,33z M21.4,15.9c0.3-0.1,0.5,0.1,0.6,0.3c0.1,0.2,0,0.5-0.3,0.6c-1.8,0.7-3.2,2.2-3.9,4
    c-0.1,0.3-0.4,0.4-0.6,0.3c-0.2-0.1-0.4-0.4-0.3-0.6C17.6,18.4,19.3,16.7,21.4,15.9z"/>
</svg>
```
:::info
If the SVG is used by itself as a contextual image, the combo of using`<svg> + role="img" + <title> + <desc> + aria-labelledby="[ID]"` provides
the most reliable choice for the different browser and screen readers that were tested.
:::

### Links
**Skip link** allows keyboard users to be able to avoid some areas of page and/or give them a more direct access to a content area,
enabling a faster navigation experience.
#### :tada: GOOD skip link example
```html{2,3,5,10}
<nav class="au-skip-link" aria-label="skip links navigation">
  <a class="au-skip-link__link" href="#content">Skip to main content</a>
  <a class="au-skip-link__link" href="#nav">Skip to main navigation</a>
</nav>
<nav id="nav" tabindex="-1" aria-label="skip links navigation">
  <ul>
    <li><a href="#">Some navigation</a></li>
  </ul>
</nav>
<div id="content" tabindex="-1">
  Some content here
</div>
```
:::info

> The content that the skip links goes to is linked with an id and it moves the focus to the element with tabindex="-1".

Reviewer: Is `tabindex` required? My understanding was it would from that point onwards interact with that element

Author: Apparently? Got the `tabindex=-1` from the `accessibility.md` 
- Check if a ‘skip link’, is it targeted via ‘#fragment’ and does the target have tabindex=”-1” applied.

and also here https://designsystem.gov.au/components/skip-link/

Reviewer: Oh neat today I learned 🎉

:::

#### :lady_beetle: BAD skip link
```html
<nav id="nav">
    <ul>
        <li><a href="#">Some navigation</a></li>
    </ul>
</nav>
<nav class="au-skip-link" aria-label="skip links navigation">
    <a class="au-skip-link__link" href="#content">Skip to content</a>
</nav>
<div id="content" tabindex="-1">
    Some content here
</div>
```
:::warning

> Skip link should be placed at the top of the page

Reviewer: ```suggestion
Because the skip link is placed after the navigation on every page load a user will have to tab through every navigation link
```

:::
**Links in general**

There's a few things to look out for when working with links:

> 1. Use `aria-label`, `aria-labelledby` or the `title` attribute to provide a more descriptive accessible name when the text content of the link cannot be changed.

Reviewer: ```suggestion
1. Use `aria-label`, `aria-labelledby` or the `title` attribute to provide a more descriptive name when the text content of the link cannot be changed.
```

> 2. All links should the focus style working

Reviewer: ```suggestion
2. All links should have focus styling (don't disable the outline without an alternative)
```

> 3. Use the right elements. Links should always be `<a>` and actions should be `<button>`

Reviewer: ```suggestion
3. Use the right elements. Links should always be `<a>` elements and actions should be `<button>`
```

#### :tada: GOOD link example
```html
<a href="post.php?post=632" aria-label="More on Using Meaningful Link Text">More...</a>
```
:::tip

> Would be even better if you could just avoid using non-informative link phrases(i.e."See more", "Click here") as link text.

Reviewer: ```suggestion
This would be even better if you could just avoid using non-informative link phrases (i.e."See more", "Click here") as link text.
```

:::

#### :lady_beetle: BAD link example
```html
<a href="post.php?post=632">
    <img src="cat.svg" aria-label="cat svg here">
    More...
</a>
```
::: warning

> If the content of a link includes an img element(icon.etc), check if the icon actually needs the aria-label, the aria-label should also be descriptive

Reviewer: ```suggestion
If the content of a link includes an `img` element (e.g. an icon), check if the `img` actually needs the aria-label, the aria-label should also be descriptive
```

:::
### Buttons
Buttons are fairly straightforward, ensure that: 

> - Check incorrect `aria-role` has been applied to any buttons.

Reviewer: I don't think we need this item as buttons will always have the same role

> - Check context has been applied to the action/button e.g. `aria-label="my contextual text".`

Reviewer: ```suggestion
- If the buttons content does not provide context then check context has been applied to the button `aria-label="my contextual text".`
```

> - `aria-controls` should be added to the button if it controls another component. (i.e. opening nav)

Reviewer: We should add a generic doc about `aria` and how it can be easily misused and some of the misconceptions (e.g. we're actually only adding `aria-controls` for the JAWS screen reader https://a11ysupport.io/tech/aria/aria-controls_attribute)

Author: Cool, will do that next. Also wow, didn't know aria-controls really is supported by JAWS screen reader only. https://a11ysupport.io/tech/aria/aria-controls_attribute this is cool 🤓

#### :tada: GOOD button example
```html
<button aria-label="open navigation" aria-controls="nav" aria-expanded="false">Menu</button>
```
:::info

> When the nav is opened, `aria-exapanded` should be true instead

Reviewer: ```suggestion
`aria-expanded` should reflect the state of the `#nav`. Ensure its updated to `true`, when the nav is opened.
```

:::
#### :lady_beetle: BAD button example
```html
<button><a href="hi.html">Click me!</a></button>
<div>I am a button!</div>
```
:::warning
`<button>` should not be used for links. Also try to avoid styling `<div>` as button, as you would need to do a lot of extra work to ensure that it's accessible(see resources button vs div)

> :::

Reviewer: ```suggestion
#### :lady_beetle: BAD button example
```html
<a class="button" onclick="runMyJs()">Click me!</a>
<div onclick="runMyJs()">I am a button!</div>
```
:::warning
In both examples we're using elements that are not `<button />` elements to trigger interactions. Screen readers will not be able to tell that the `a` tag is not a link and they will not be able to tell that the `div` is intractable
:::
```

### Navigation

> Important thing to remember for navigation is that we should also ensure that keyboard shortcuts work for controlling the menu e.g. `esc` key `closes`, `space` key `opens` etc.

Reviewer: ```suggestion
When implementing navigation through modal interfaces it's important to remember that you should implement the expected controls:
- `esc` closes the modal
- `tab` will cycle through only the items inside the modal
- `shift + tab` will cycle through in reverse only the items inside the modal
```

If a button controls the open state of the navigation, the button should have `aria-expanded` attribute (see Button example above)
#### :tada: GOOD Navigation example
```vue{1,6}
<nav id="nav" role="navigation" aria-label="main">
    <ul>
      <li><a href="#">Some navigation</a></li>  
    </ul>
</nav>
<secondary-nav id="secondary-nav" role="navigation" aria-label="secondary">
    <ul>
      <li><a href="#">Some navigation</a></li>
    </ul>
</seconadry-nav>
```
:::tip

> We should use `aria-label` to help user to distinguish the different nav available, when using `aria-label` on a <a href="https://web.dev/use-landmarks/">landmark element(i.e. nav)</a>, don't include the word role of the landmark, as screen readers will annouce it afterwards.

Reviewer: ```suggestion
We should use `aria-label` to help user to distinguish the different `nav` elements available, when using `aria-label` on a <a href="https://web.dev/use-landmarks/">landmark element(i.e. nav)</a>, don't include the word role of the landmark, as screen readers will annouce it afterwards.
```

:::

#### :lady_beetle: BAD Navigation example
```html
<nav id="nav">
    <ul>
        <li><a href="#">Some navigation</a></li>
    </ul>
</nav>
```
:::info

> `role=”navigation”` is not required on explicit tags like `<nav>`. However older screen readers that do not support HTML5 tags will require it.

Reviewer: If thats the case could we axe this example? I think we shouldn't need to worry about old browsers in this regard
